Objective-C/CBLDatabase.mm

```diff
@@ -864,139 +870,201 @@ - (CBLDocument*) saveDocument: (CBLDocument*)document
+                                                             base: baseDoc];
+        CBLDocument* resolved = [resolver resolve: conflict];
+        if (!resolved) {
+            convertError({LiteCoreDomain, kC4ErrorConflict}, outError);
             return nil;
-        Assert(newDoc); // In a transaction we can't have a second conflict after merging!
-    }
-    
-    // Save succeeded; now commit the transaction:
-    if (!transaction.commit()) {
-        c4doc_free(newDoc);
-        convertError(transaction.error(), outError);
-        return nil;
+        }
+        
+        CBL_LOCK(self) {
+            CBLDocument* current = [[CBLDocument alloc] initWithDatabase: self
+                                                              documentID: docID
+                                                          includeDeleted: YES
+                                                                   error: outError];
+            if ($equal(resolved.revID, current.revID))
+                return resolved; // Same as current
+            
+            // For saving:
+            doc = resolved;
+            baseRev = current.c4Doc.rawDoc;
+            deletion = resolved.isDeleted;
+        }
     }
-    
-    CBLC4Document* c4Doc = [CBLC4Document document: newDoc];
-    return [[CBLDocument alloc] initWithDatabase: self
-                                      documentID: document.id
-                                           c4Doc: c4Doc];
 }
 
 
 #pragma mark - RESOLVING REPLICATED CONFLICTS:
 
 
+// Called inside the replicator's database transaction.
 - (bool) resolveConflictInDocument: (NSString*)docID
                      usingResolver: (id<CBLConflictResolver>)resolver
-                             error: (NSError**)outError {
-    C4Transaction t(_c4db);
-    t.begin();
-
-    auto doc = [[CBLDocument alloc] initWithDatabase: self
-                                          documentID: docID
-                                           mustExist: YES
-                                      includeDeleted: YES
-                                               error: outError];
-    if (!doc)
-        return false;
-
-    // Read the conflicting remote revision:
-    auto otherDoc = [[CBLDocument alloc] initWithDatabase: self
-                                               documentID: docID
-                                                mustExist: YES
-                                           includeDeleted: YES
-                                                    error: outError];
-    if (!otherDoc || ![otherDoc selectConflictingRevision])
-        return false;
-
-    // Read the common ancestor revision (if it's available):
-    auto baseDoc = [[CBLDocument alloc] initWithDatabase: self
+                             error: (NSError**)outError
+{
+    CBLDocument* doc, *otherDoc, *baseDoc;
+    CBL_LOCK(self) {
+        // Open transaction to be able to see current uncommitted changes
```

Comment:
What does this mean? Opening a transaction doesn't let you see changes made by other transactions.